### ISSUE TITLE
panel crashes on SIGTERM

### ISSUE THREAD

**Issue by RalfJung** [lxde/lxqt#994]
lxqt-panel crashes when it receives a SIGTERM:
```
#0  0x0000000000000000 in ?? ()
#1  0x00000000004b1d94 in qDeleteAll<QList<TrayIcon*>::const_iterator> (begin=..., end=...) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qalgorithms.h:317
#2  0x00000000004b19f6 in qDeleteAll<QList<TrayIcon*> > (c=...) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qalgorithms.h:325
#3  0x00000000004b131c in LXQtTray::stopTray (this=0x8ab340) at /home/r/src/lxqt/lxqt-panel/plugin-tray/lxqttray.cpp:363
#4  0x00000000004b07ca in LXQtTray::~LXQtTray (this=0x8ab340, __in_chrg=<optimized out>) at /home/r/src/lxqt/lxqt-panel/plugin-tray/lxqttray.cpp:93
#5  0x00000000004b0838 in LXQtTray::~LXQtTray (this=0x8ab340, __in_chrg=<optimized out>) at /home/r/src/lxqt/lxqt-panel/plugin-tray/lxqttray.cpp:94
#6  0x00000000004b0560 in LXQtTrayPlugin::~LXQtTrayPlugin (this=0x8bda30, __in_chrg=<optimized out>) at /home/r/src/lxqt/lxqt-panel/plugin-tray/lxqttrayplugin.cpp:41
#7  0x00000000004b05b4 in LXQtTrayPlugin::~LXQtTrayPlugin (this=0x8bda30, __in_chrg=<optimized out>) at /home/r/src/lxqt/lxqt-panel/plugin-tray/lxqttrayplugin.cpp:42
#8  0x0000000000450d1a in Plugin::~Plugin (this=0x87bf70, __in_chrg=<optimized out>) at /home/r/src/lxqt/lxqt-panel/panel/plugin.cpp:168
#9  0x0000000000450dc8 in Plugin::~Plugin (this=0x87bf70, __in_chrg=<optimized out>) at /home/r/src/lxqt/lxqt-panel/panel/plugin.cpp:174
#10 0x000000000043df76 in qDeleteAll<QList<Plugin*>::const_iterator> (begin=..., end=...) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qalgorithms.h:317
#11 0x000000000043c9c8 in qDeleteAll<QList<Plugin*> > (c=...) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qalgorithms.h:325
#12 0x0000000000438611 in PanelPluginsModel::~PanelPluginsModel (this=0x7b5de0, __in_chrg=<optimized out>) at /home/r/src/lxqt/lxqt-panel/panel/panelpluginsmodel.cpp:50
#13 0x0000000000438678 in PanelPluginsModel::~PanelPluginsModel (this=0x7b5de0, __in_chrg=<optimized out>) at /home/r/src/lxqt/lxqt-panel/panel/panelpluginsmodel.cpp:51
#14 0x0000000000446d62 in QScopedPointerDeleter<PanelPluginsModel>::cleanup (pointer=0x7b5de0) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qscopedpointer.h:54
#15 0x0000000000446535 in QScopedPointer<PanelPluginsModel, QScopedPointerDeleter<PanelPluginsModel> >::~QScopedPointer (this=0x7b1e18, __in_chrg=<optimized out>) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qscopedpointer.h:101
#16 0x0000000000441e59 in LXQtPanel::~LXQtPanel (this=0x7b1dc0, __in_chrg=<optimized out>) at /home/r/src/lxqt/lxqt-panel/panel/lxqtpanel.cpp:316
#17 0x0000000000441eae in LXQtPanel::~LXQtPanel (this=0x7b1dc0, __in_chrg=<optimized out>) at /home/r/src/lxqt/lxqt-panel/panel/lxqtpanel.cpp:320
#18 0x0000000000449c01 in qDeleteAll<QList<LXQtPanel*>::const_iterator> (begin=..., end=...) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qalgorithms.h:317
#19 0x00000000004493bd in qDeleteAll<QList<LXQtPanel*> > (c=...) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qalgorithms.h:325
#20 0x0000000000448474 in LXQtPanelApplication::cleanup (this=0x7fffffffdf50) at /home/r/src/lxqt/lxqt-panel/panel/lxqtpanelapplication.cpp:93
#21 0x000000000044aebe in QtPrivate::FunctorCall<QtPrivate::IndexesList<>, QtPrivate::List<>, void, void (LXQtPanelApplication::*)()>::call(void (LXQtPanelApplication::*)(), LXQtPanelApplication*, void**) (
    f=(void (LXQtPanelApplication::*)(LXQtPanelApplication * const)) 0x448458 <LXQtPanelApplication::cleanup()>, o=0x7fffffffdf50, arg=0x7fffffffde80) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs_impl.h:501
#22 0x000000000044aa10 in QtPrivate::FunctionPointer<void (LXQtPanelApplication::*)()>::call<QtPrivate::List<>, void>(void (LXQtPanelApplication::*)(), LXQtPanelApplication*, void**) (
    f=(void (LXQtPanelApplication::*)(LXQtPanelApplication * const)) 0x448458 <LXQtPanelApplication::cleanup()>, o=0x7fffffffdf50, arg=0x7fffffffde80) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qobjectdefs_impl.h:520
#23 0x000000000044a407 in QtPrivate::QSlotObject<void (LXQtPanelApplication::*)(), QtPrivate::List<>, void>::impl(int, QtPrivate::QSlotObjectBase*, QObject*, void**, bool*) (which=1, this_=0x7b1ce0, r=0x7fffffffdf50, a=0x7fffffffde80, 
    ret=0x0) at /usr/include/x86_64-linux-gnu/qt5/QtCore/qobject_impl.h:143
#24 0x00007ffff41cfe0f in QtPrivate::QSlotObjectBase::call (a=0x7fffffffde80, r=0x7fffffffdf50, this=<optimized out>) at ../../include/QtCore/../../src/corelib/kernel/qobject_impl.h:124
#25 QMetaObject::activate (sender=0x7fffffffdf50, signalOffset=<optimized out>, local_signal_index=local_signal_index@entry=0, argv=argv@entry=0x0) at kernel/qobject.cpp:3698
#26 0x00007ffff41d0797 in QMetaObject::activate (sender=<optimized out>, m=m@entry=0x7ffff43ea5e0 <QCoreApplication::staticMetaObject>, local_signal_index=local_signal_index@entry=0, argv=argv@entry=0x0) at kernel/qobject.cpp:3578
#27 0x00007ffff424eea0 in QCoreApplication::aboutToQuit (this=<optimized out>) at .moc/moc_qcoreapplication.cpp:242
#28 0x00007ffff41a6e83 in QCoreApplication::exec () at kernel/qcoreapplication.cpp:1234
#29 0x0000000000438532 in main (argc=1, argv=0x7fffffffe068) at /home/r/src/lxqt/lxqt-panel/panel/main.cpp:41
```

This is happening only with lxqt-panel/master. I don't see a crash with the Debian packaged version (0.10).

**Comment by palinek** [comment]
I'm not able to reproduce it here...

@RalfJung can you, please, check if the lxde/lxqt-panel#313 fixes this for you.

**Comment by paulolieuthier** [comment]
I was trying to fix this yesterday, but I'm getting a segfault from KWindowSystem instead:

```
Thread 1 "lxqt-panel" received signal SIGTERM, Terminated.
0x00007ffff1b8303f in pthread_cond_wait@@GLIBC_2.3.2 ()
   from /usr/lib/libpthread.so.0
(gdb) bt
#0  0x00007ffff1b8303f in pthread_cond_wait@@GLIBC_2.3.2 ()
   from /usr/lib/libpthread.so.0
#1  0x00007ffff56add39 in ?? () from /usr/lib/libxcb.so.1
#2  0x00007ffff56af497 in ?? () from /usr/lib/libxcb.so.1
#3  0x00007ffff56af5a1 in xcb_wait_for_reply () from /usr/lib/libxcb.so.1
#4  0x00007ffff4b1da9a in ?? () from /usr/lib/libKF5WindowSystem.so.5
#5  0x00007ffff4b19eab in NETWinInfo::update(QFlags<NET::Property>, QFlags<NET::Property2>) () from /usr/lib/libKF5WindowSystem.so.5
#6  0x00007ffff4b1c6c7 in NETWinInfo::event(xcb_generic_event_t*, QFlags<NET::Property>*, QFlags<NET::Property2>*) () from /usr/lib/libKF5WindowSystem.so.5
#7  0x00007fffd857a265 in ?? ()
   from /usr/lib/qt/plugins/kf5/org.kde.kwindowsystem.platforms/KF5WindowSystemX11Plugin.so
#8  0x00007ffff348d84f in QAbstractEventDispatcher::filterNativeEvent(QByteArray const&, void*, long*) () from /usr/lib/libQt5Core.so.5
#9  0x00007fffea7c89d5 in QXcbConnection::handleXcbEvent(xcb_generic_event_t*) ()
   from /usr/lib/libQt5XcbQpa.so.5
#10 0x00007fffea7c9906 in QXcbConnection::processXcbEvents() ()
   from /usr/lib/libQt5XcbQpa.so.5
#11 0x00007ffff34bd659 in QObject::event(QEvent*) ()
   from /usr/lib/libQt5Core.so.5
#12 0x00007ffff43b868c in QApplicationPrivate::notify_helper(QObject*, QEvent*)
    () from /usr/lib/libQt5Widgets.so.5
#13 0x00007ffff43bd8f6 in QApplication::notify(QObject*, QEvent*) ()
   from /usr/lib/libQt5Widgets.so.5
#14 0x00007ffff3490518 in QCoreApplication::notifyInternal2(QObject*, QEvent*) ()
   from /usr/lib/libQt5Core.so.5
#15 0x00007ffff34923da in QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) () from /usr/lib/libQt5Core.so.5
#16 0x00007ffff34e5883 in ?? () from /usr/lib/libQt5Core.so.5
#17 0x00007ffff7706d87 in g_main_context_dispatch ()
   from /usr/lib/libglib-2.0.so.0
#18 0x00007ffff7706fe0 in ?? () from /usr/lib/libglib-2.0.so.0
#19 0x00007ffff770708c in g_main_context_iteration ()
   from /usr/lib/libglib-2.0.so.0
#20 0x00007ffff34e5c8f in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/libQt5Core.so.5
#21 0x00007ffff348e6fa in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>)
    () from /usr/lib/libQt5Core.so.5
#22 0x00007ffff3496b7c in QCoreApplication::exec() ()
   from /usr/lib/libQt5Core.so.5
#23 0x00000000004938f3 in main (argc=1, argv=0x7fffffffe108)
    at /home/paulo/Workspace/lxqt/lxqt-panel/panel/main.cpp:41
```

**Comment by paulolieuthier** [comment]
And if I remove the taskbar plugin:

```
Thread 1 "lxqt-panel" received signal SIGTERM, Terminated.
0x00007ffff26b2c3d in poll () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007ffff26b2c3d in poll () from /usr/lib/libc.so.6
#1  0x00007ffff7706f7c in ?? () from /usr/lib/libglib-2.0.so.0
#2  0x00007ffff770708c in g_main_context_iteration ()
   from /usr/lib/libglib-2.0.so.0
#3  0x00007ffff34e5c8f in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/libQt5Core.so.5
#4  0x00007ffff348e6fa in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>)
    () from /usr/lib/libQt5Core.so.5
#5  0x00007ffff3496b7c in QCoreApplication::exec() ()
   from /usr/lib/libQt5Core.so.5
#6  0x00000000004938f3 in main (argc=1, argv=0x7fffffffe108)
    at /home/paulo/Workspace/lxqt/lxqt-panel/panel/main.cpp:41
```

**Comment by RalfJung** [comment]
This doesn't look like a SIGSEV, it looks like the SIGTERM that you sent to the application? It's kind of expected that we see those somewhere where the application blocks. I get that, too, of course, then I press "c" to continue executing the application, and *then* I see a SIGSEV.

**Comment by paulolieuthier** [comment]
I get the same results with SIGSEGV.

**Comment by pmattern** [comment]
In qps there's a [crash](https://github.com/QtDesktop/qps/issues/13) the traces of which look a bit similar to the ones of the problem discussed here.
Those crashes seems to be related to Qt 5.*6* as they got manifest only after compiling the latest commit of qps against this version but not during the roughly four weeks I was running the very same commit compiled against Qt 5.*5*.

Btw. I can confirm the crash discussed here as well, traces on https://gist.github.com/pmattern/ca2665d426cd286cffd0cba0357cdaf3.

**Comment by RalfJung** [comment]
I am using Qt 5.5 both for compiling lxqt-panel and for running it.

**Comment by tsujan** [comment]
No crash for me, with Debian Testing and LXQT compiled from the git sources.

@RalfJung Do you have a 32-bit tray icon like Skype's? Which tray icons do you usually have?

**Comment by RalfJung** [comment]
In terms of "old-style" tray icons, I have nm-applet, Thunderbird (via FireTray), Gajim and sometimes Liferea.

**Comment by tsujan** [comment]
It would be informative if you test without them, one by one.

**Comment by RalfJung** [comment]
Actually, I see crashes even with no tray item.

Maybe it is related to the way I test this -- I log into a session which spawns the lxqt-panel provided by Debian (which does not crash), and then I go to the build directory of the panel and do
```
killall lxqt-panel ; sleep 0.5 ; panel/lxqt-panel  &
```
After some seconds, I will do that again. Then the first panel/lxqt-panel crashes.

**Comment by tsujan** [comment]
`killall lxqt-panel ; sleep 0.5 ; lxqt-panel` doesn't cause any crash here.

**Comment by RalfJung** [comment]
I am getting the impression that the crash is caused by mixing libraries compiled from git master (installed to `/opt/lxqt`), and the system-wide LXQt installation in `/usr`. Is there documentation anywhere which environment variables need to be set for this to work?

**Comment by pmattern** [comment]
I'm seeing the crashes on Arch systems using each and every LXQt component from recent VCS checkouts.

**Comment by tsujan** [comment]
> I'm seeing the crashes on Arch systems using each and every LXQt component from recent VCS checkouts.

That should be something else. This is about a crash that, until now, only @RalfJung can reproduce, @paulolieuthier's case being different.

**Comment by pmattern** [comment]
@tsujan 
Well, I cannot reproduce the crash by terminating with `kill`, no matter which signal I use. But the panel is crashing upon exiting sessions all the time here, both "bare metal" and in virtual machines.

**Comment by tsujan** [comment]
@pmattern 
Any crash log, coredump?

You said every LXQT component crashes under Arch. That may be related to a core LXQT component then.

I know Manjaro isn't Arch but its Qt is updated to v5.6 and there's no crash. However, there, I haven't compiled LXQT from the latest sources. I'll do it when I have time.

**Comment by tsujan** [comment]
Oh, I was wrong! After the Qt5 upgrade under Manjaro, I have an lxqt-panel coredump with `Signal: 11 (SEGV)` and this uninformative backtrace:
````
#0  0x00007f238d656ab7 in QObject::disconnect(QObject const*, char const*, QObject const*, char const*) ()
   from /usr/lib/libQt5Core.so.5
#1  0x00007f238eee9f78 in ?? () from /usr/lib/libQt5DBus.so.5
#2  0x00007f238eedab36 in ?? () from /usr/lib/libQt5DBus.so.5
#3  0x00007f238d452de8 in ?? () from /usr/lib/libQt5Core.so.5
#4  0x00007f238bf29424 in start_thread () from /usr/lib/libpthread.so.0
#5  0x00007f238ca67cbd in clone () from /usr/lib/libc.so.6
````
But I also have many other coredumps, all of which, except for the above one, are related to the KDE session (akonadi, kalarm, kdeconnectd, kded5 and all user startup programs, including Qt5, non-KDE ones)! I tested some of them with `kill` and, as @pmattern said, no crash.

Such a mess didn't exist before the Qt upgrade. This may go beyond LXQT. My DM is sddm under Manjaro.

**Comment by RalfJung** [comment]
I still have Qt 5.5, and I don't see anything KDE-related.

However, as I already mentioned, I have both LXQt 0.10 and LXQt master installed, and I suspect it is mixing up the two. Which environment variables do I need to set to prefer the LXQt installed in `/opt/lxqt` over the distro-installed one in `/usr`? I have set `LD_LIBRARY_PATH`, what esle? `QT_PLUGIN_PATH` maybe, and is there a separate environment variable for LXQt plugins?

**Comment by tsujan** [comment]
@RalfJung 
Unfortunately, because of the title of this issue, two completely different problems are mixed together here. The first one is your crash with the specific setting you have (btw, I don't know the answer to your question); the second one is about Qt-5.6 + Arch. Under Debian Testing (my work system) and with the latest component compiled from git sources, I see no crash at all. Under an updated Manjaro and with LXQT from its repository, I see what I described, which is like what @pmattern sees under Arch with the latest LXQT.

**Comment by palinek** [comment]
I think, this issue (the @RalfJung's original) should be closed as fixed by lxde/lxqt-panel#313 and another one opened for the Qt5.6 problem:
> I'm seeing the crashes on Arch systems using each and every LXQt component from recent VCS checkouts.

**Comment by tsujan** [comment]
> I think, this issue (the @RalfJung's original) should be closed as fixed by lxde/lxqt-panel#313 and another one opened for the Qt5.6 problem

I agree, especially when the latter may be a downstream issue.

**Comment by tsujan** [comment]
Closed. Please open another issue if Arch/Manjaro problem persists!
